### PR TITLE
[core] Share channel with GcsClient.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
According to https://grpc.io/docs/guides/performance/ we should: Always re-use stubs and channels when possible.

This PR share channels between different services.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/rpc/grpc_client.h [hunk 1]
```diff
@@ -45,22 +45,59 @@ namespace rpc {
...
+}
+
 template <class GrpcService>
 class GrpcClient {
  public:
+  GrpcClient(std::shared_ptr<grpc::Channel> channel,
+             ClientCallManager &call_manager,
+             bool use_tls = false)
+      : client_call_manager_(call_manager), use_tls_(use_tls) {
+    stub_ = GrpcService::NewStub(channel);
```
Comment (mwtian) [inline src/ray/rpc/grpc_client.h:92]: `std::move(channel)`?

File: src/ray/rpc/grpc_client.h [hunk 2]
```diff
@@ -45,22 +45,59 @@ namespace rpc {
...
-    // Disable http proxy since it disrupts local connections. TODO(ekl) we should make
-    // this configurable, or selectively set it for known local connections only.
-    argument.SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 0);
-    argument.SetMaxSendMessageSize(::RayConfig::instance().max_grpc_message_size());
-    argument.SetMaxReceiveMessageSize(::RayConfig::instance().max_grpc_message_size());
-
-    std::shared_ptr<grpc::Channel> channel = BuildChannel(argument, address, port);
+    std::shared_ptr<grpc::Channel> channel = BuildChannel(address, port);
 
     stub_ = GrpcService::NewStub(channel);
```
Comment (mwtian) [inline src/ray/rpc/grpc_client.h:102]: `std::move(channel)`?

**Review by mwtian — APPROVED** [review]

**Comment by scv119** [comment]
will we expect longer queuing time since now all the services are on the same channel?

**Comment by iycheng** [comment]
@scv119  I don't think so. It's a TCP connection and the bottleneck should be in the network layer which is the same.

**Comment by iycheng** [comment]
@mwtian  I'll fix the comment in the next update. Let the test run for a while first.

**Review by scv119 — APPROVED** [review]

**Comment by mwtian** [comment]
Also I believe gRPC uses the same underlying connection for channels with the same args.

**Comment by iycheng** [comment]
Release test failed for test_many_tasks. I'm investigating. Many actor seems faster.. hmm.. I'll need to observe more for this before merging.

**Comment by iycheng** [comment]
> Also I believe gRPC uses the same underlying connection for channels with the same args.

I mannually checked it by printing the address of the Channels. They have different addresses. But I also checked the TCP connections by using lsof and only one TCP connection therefrom rayelt -> gcs.

**Comment by iycheng** [comment]
It seems ok once merging the master.

https://console.anyscale.com/o/anyscale-internal/projects/prj_LFKjNprpiYt4AjAf1NdLDJwn/clusters/ses_aRFPL5JEpDZHNLp71UGJ9Ffp?command-history-section=command_history

**Pull request merged**